checks/shell_download_validate.go

```diff
@@ -0,0 +1,628 @@
+		if !ok {
+			continue
+		}
+		if len(v.Parts) != 1 {
+			continue
+		}
+
+		lit, ok := v.Parts[0].(*syntax.Lit)
+		if !ok {
+			continue
+		}
+		return lit.Value, true
+	}
+	return "", false
+}
+
+func validateCommandIsNotNotSuFetchToStdinExecute(cmd, pathfn string,
+	logf func(s string, f ...interface{})) (bool, error) {
+	in := strings.NewReader(cmd)
+	f, err := syntax.NewParser().Parse(in, "")
+	if err != nil {
+		return false, ErrParsingShellCommand
+	}
+
+	cmdValidated := true
+	syntax.Walk(f, func(node syntax.Node) bool {
+		if isFetchStdinExecute(node, cmd, pathfn, logf) {
+			cmdValidated = false
+		}
+
+		// Continue walking the node graph.
+		return true
+	})
+
+	return cmdValidated, nil
+}
+
+func isSuFetchStdinExecute(node syntax.Node, cmd, pathfn string,
+	logf func(s string, f ...interface{})) (bool, error) {
+	ce, ok := node.(*syntax.CallExpr)
+	if !ok {
+		return false, nil
+	}
+
+	c, ok := extractCommand(ce)
+	if !ok {
+		return false, nil
+	}
+
+	if !isSuCommand(c) {
+		return false, nil
+	}
+
+	cs, ok := extractSuCommand(ce.Args)
+	if !ok {
+		return false, nil
+	}
+
+	// We have extracted the command. Now parse it.
+	in := strings.NewReader(cs)
+	f, err := syntax.NewParser().Parse(in, "")
+	if err != nil {
+		return false, fmt.Errorf("syntax.NewParser: %w", err)
+	}
+
+	isFetch := false
+	syntax.Walk(f, func(node syntax.Node) bool {
+		if isFetchStdinExecute(node, cmd, pathfn, logf) {
+			isFetch = true
+		}
+
+		// Continue walking the node graph.
+		return true
+	})
+
+	return isFetch, nil
+}
+
+func validateCommandIsNotSuFetchToStdinExecute(cmd, pathfn string,
+	logf func(s string, f ...interface{})) (bool, error) {
+	in := strings.NewReader(cmd)
+	f, err := syntax.NewParser().Parse(in, "")
+	if err != nil {
+		return false, ErrParsingShellCommand
+	}
+
+	cmdValidated := true
+	syntax.Walk(f, func(node syntax.Node) bool {
+		if err != nil {
+			return false
+		}
+		var b bool
+		b, err = isSuFetchStdinExecute(node, cmd, pathfn, logf)
+		if err != nil {
+			return false
+		} else if b {
+			cmdValidated = false
+		}
+
+		// Continue walking the node graph.
+		return true
+	})
+
+	return cmdValidated, err
+}
+
+func validateCommandIsNotFetchToStdinExecute(cmd, pathfn string,
+	logf func(s string, f ...interface{})) (bool, error) {
+	r1, err := validateCommandIsNotNotSuFetchToStdinExecute(cmd, pathfn, logf)
+	if err != nil {
+		return r1, err
+	}
+
+	r2, err := validateCommandIsNotSuFetchToStdinExecute(cmd, pathfn, logf)
+	if err != nil {
+		return r2, err
+	}
+
+	return r1 && r2, nil
+}
+
+func recordFetchFileFromString(cmd string) (fmap map[string]bool, ok bool, err error) {
+	in := strings.NewReader(cmd)
+	f, err := syntax.NewParser().Parse(in, "")
+	if err != nil {
+		return nil, false, ErrParsingShellCommand
+	}
+
+	files := make(map[string]bool)
+	syntax.Walk(f, func(node syntax.Node) bool {
+		if err != nil {
+			return false
+		}
+		var b bool
+		var fn string
+		fn, b, err = recordFetchFileFromNode(node)
```

Comment:
can just be
```fn, b, err := recordFetchFileFromNode(node)```

Go lets you do this even though `err` was defined previously. Same in other places (e.g. line 550)

Reply:
what will be the output of this? Will `err` be an alias to the global `err` previously defined or is it a local variable with the same name? (Note: I've changed this in a. follow-up-CL to look like `fn, b, e := recordFetchFileFromNode(node)`)

Reply:
per https://golang.org/ref/spec#Short_variable_declarations

"Redeclaration does not introduce a new variable; it just assigns a new value to the original."

Reply:
ah actually in this case, because you're in another block the scope is different and you'll create a new one instead.